OpenRA.Mods.Common/Traits/Buildings/Building.cs

```diff
@@ -177,6 +176,9 @@ public Actor FindBaseProvider(World world, Player p, CPos topLeft)
 
 		public virtual bool IsCloseEnoughToBase(World world, Player p, string buildingName, CPos topLeft)
 		{
+			Func<Actor, bool> hasAdjanentType = a => a.Info.TraitInfoOrDefault<GivesBuildableAreaInfo>().ValidTypes.Contains(AdjacentType);
+			Func<Actor, bool> givesBuildableAreaNotNull = a => a.Info.TraitInfoOrDefault<GivesBuildableAreaInfo>() != null;
```

Comment:
I think @pchote had one combined function in mind, like:
```
Func<Actor, bool> hasAdjanentType = a =>
{
	var givesBuildableArea = a.Info.TraitInfoOrDefault<GivesBuildableAreaInfo>();
	if (givesBuildableArea == null)
		return false;

	return givesBuildableArea.ValidTypes.Count == 0 || givesBuildableArea.ValidType.Contains(AdjacentType);
};
```
(Note: I also added a `Count == 0` to ensure that `Leave empty to accept all types.` works. Did that work before without that check?)